libraries/I2S/src/I2S.cpp

```diff
@@ -0,0 +1,209 @@
+/*
+  Based off of ArduinoCore-SAMD I2S interface.  Modified for the
+  ESP8266 by Earle F. Philhower, III <earlephilhower@yahoo.com>
+
+  This library is free software; you can redistribute it and/or
+  modify it under the terms of the GNU Lesser General Public
+  License as published by the Free Software Foundation; either
+  version 2.1 of the License, or (at your option) any later version.
+
+  This library is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+  See the GNU Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public
+  License along with this library; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+*/
+
+#include <Arduino.h>
+#include "I2S.h"
+
+I2SClass::I2SClass(bool enableTransmit, bool enableRecv, bool driveClocks) {
+  _enableTx = enableTransmit;
+  _enableRx = enableRecv;
+  _driveClk = driveClocks;
+  _running = false;
+  _onTransmit = nullptr;
+  _onReceive = nullptr;
+  _havePeeked = 0;
+  _peekedData = 0;
+  _bps = 0;
+  _writtenHalf = false;
+}
+
+int I2SClass::begin(int mode, long sampleRate, int bitsPerSample) {
+  if ( _running || (mode != I2S_PHILIPS_MODE) || ( (bitsPerSample != 16) && (bitsPerSample != 24) ) ) {
+    return 0;
+  }
+  if (!i2s_rxtxdrive_begin(_enableRx, _enableTx, _driveClk, _driveClk)) {
+    return 0;
+  }
+  i2s_set_rate(sampleRate);
+  i2s_set_callback(_onTransmit);
+  i2s_rx_set_callback(_onReceive);
+  _bps = bitsPerSample;
+  _running = true;
+  return 1;
+}
+
+void I2SClass::end() {
+  if (_running) {
+    i2s_end();
+  }
+  i2s_set_callback(nullptr);
+  i2s_rx_set_callback(nullptr);
+  _running = false;
+}
+
+void I2SClass::onTransmit(void(*fcn)(void)) {
+  i2s_set_callback(fcn);
+  _onTransmit = fcn;
+}
+
+void I2SClass::onReceive(void(*fcn)(void)) {
+  i2s_rx_set_callback(fcn);
+  _onReceive = fcn;
+}
+
+int I2SClass::available() {
+  if (!_running) return 0;
+  return i2s_rx_available();
+}
+
+int I2SClass::availableForWrite() {
+  if (!_running) return 0;
+  return i2s_available();
+}
+
+void I2SClass::flush() {
+  /* No-op */
+}
+
+int I2SClass::read() {
+  if (!_running) return -1;
+  if (_havePeeked) {
```

Comment:
To avoid recursion, would this be better, or again is it preferable to keep sync with upstream lib ?
```
if (!_havePeeked) peek();
if (!_havePeeked) return -1;
if (_bps ==16) ...
```

Reply:
Good idea, done.